### PR TITLE
[Sensor] Fix hab-lab examples to work with Hierarchical SensorSpec changes

### PR DESCRIPTION
**Pull request by vauduong**
## Motivation and Context

See: https://github.com/facebookresearch/habitat-sim/pull/1090
PR#1090 in habitat-sim introduces a Hierarchical SensorSpec to match the new sensor framework for VisualSensor. 

image

`SensorSpec::create()` is no longer used, use correct SensorSpec initialization to create correct sensor, such as `CameraSensorSpec::create()`. 

This PR fixes examples to use CameraSensorSpec() to initialize sensorSpecs

## How Has This Been Tested

Built and tested locally with version of hab-sim in PR#1090, checked that tests passed

## Types of changes

- Docs change / refactoring / dependency upgrade
- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def create_sim_config(
 
         sensor_specifications = []
         for sensor in _sensor_suite.sensors.values():
-            sim_sensor_cfg = habitat_sim.SensorSpec()
+            sim_sensor_cfg = habitat_sim.CameraSensorSpec()
```
Comment (vauduong) [inline habitat/sims/habitat_simulator/habitat_simulator.py:261]: Hi @Skylion007 ! I noticed that in hab-lab sensors we don't have SensorSubTypes, and was wondering what the reasoning for that was - would it be a good idea to add this enum?

For context, I check that SensorSubTypes are correct before initializing a CameraSensorSpec() here: 
https://github.com/facebookresearch/habitat-sim/pull/1090/files#diff-1c82f32ca8e8fe4738a140dd923ac12f5d482300e28446788e41eafa248a62a1R66-R103

Comment (vauduong) [reply habitat/sims/habitat_simulator/habitat_simulator.py:261]: I'm also not sure how BumpSensor is initialized, since this isn't something I was able to find in hab-sim or hab-lab. I know we can't initialize it with CameraSensorSpec as we do the other sensors though. Any insight to this would be appreciated! @Skylion007 @bigbike @mathfac @erikwijmans @abhiskk

**Comment by Skylion007** [comment]
I started doing that here: https://github.com/facebookresearch/habitat-lab/pull/533

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 1 above)
Comment (Skylion007) [reply habitat/sims/habitat_simulator/habitat_simulator.py:261]: We don't actually have implementations of BumpSensor, I don't believe.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 2]
```diff
@@ -257,41 +263,69 @@ def create_sim_config(
...
-            # We know that the Sensor has to be one of these Sensors
-            sensor = cast(HabitatSimVizSensors, sensor)
-            sim_sensor_cfg.sensor_type = sensor.sim_sensor_type
-            sim_sensor_cfg.gpu2gpu_transfer = (
-                self.habitat_config.HABITAT_SIM_V0.GPU_GPU
-            )
-            sensor_specifications.append(sim_sensor_cfg)
+
+            # Check if type VisualSensorSpec, we know that Sensor is one of HabitatSimRGBSensor, HabitatSimDepthSensor, HabitatSimSemanticSensor
+            if (
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:278]: Not all sensors are VisualSensor in Habitat.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 3]
```diff
@@ -160,9 +164,11 @@ def get_observation(
 @registry.register_sensor
 class HabitatSimSemanticSensor(SemanticSensor):
     sim_sensor_type: habitat_sim.SensorType
+    sim_sensor_subtype: habitat_sim.SensorSubType
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:167]: I suppose, we could tie the SensorSubtype to the Sensor Class like this, but I was thinking about having it be a SensorConfig string that determined it (similar to how we do for UUIDs now). Any reason why not do it that way? It would give the user more flexability down the road (unless you expect SensorSubtypes to have entirely different config types).

Doing it like this though causes a quadratic explosion in the number of Sensor modalities -> sensor_subtypes though.

In other words, is there a good reason this cannot/should not be read from the config?

This is all fine for now since we only support PINHOLE, but will be annoying later when we supports RGBFisheye/DepthFisheye etc...

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 4]
```diff
@@ -257,41 +263,69 @@ def create_sim_config(
         )
 
         sensor_specifications = []
+        VisualSensorTypeSet = {
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:266]: We can programatically get the values from a enum by iterating through them with `.value()` 

Also wouldn't this be better defined by some inheritance structure?

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 5]
```diff
@@ -257,41 +263,69 @@ def create_sim_config(
...
+
+            # Check if type VisualSensorSpec, we know that Sensor is one of HabitatSimRGBSensor, HabitatSimDepthSensor, HabitatSimSemanticSensor
+            if (
+                getattr(sensor, "sim_sensor_type", [])
+                not in VisualSensorTypeSet
+            ):
+                raise ValueError(
+                    f"""{getattr(sensor, "sim_sensor_type", [])} is an illegal sensorType that is not implemented yet"""
+                )
+            else:
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:285]: No reason for an else statement here, athe ValueError will terminate it.

File: habitat/sims/habitat_simulator/habitat_simulator.py [hunk 6]
```diff
@@ -257,41 +263,69 @@ def create_sim_config(
...
+                # Check if type CameraSensorSpec
+                if (
+                    getattr(sensor, "sim_sensor_subtype", [])
+                    not in CameraSensorSubTypeSet
+                ):
+                    raise ValueError(
+                        f"""{getattr(sensor, "sim_sensor_subtype", [])} is an illegal sensorSubType for a VisualSensor"""
+                    )
+                else:
```
Comment (Skylion007) [inline habitat/sims/habitat_simulator/habitat_simulator.py:294]: Like wise, no reason for else

**Review by Skylion007 — CHANGES_REQUESTED** [review]
This is fine as a cludge, but there are some issues with support multiple sensor types n the future. Also, some issues with the current error handling / indentation

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 2 above)
Comment (vauduong) [reply habitat/sims/habitat_simulator/habitat_simulator.py:278]: Not all sensors are VisualSensors but VisualSensorSpecs are the only kinds of specs that can be initialized at the moment.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 3 above)
Comment (vauduong) [reply habitat/sims/habitat_simulator/habitat_simulator.py:167]: Hi Aaron! I did this because this file goes through some defined configs and initializes the right sensorSpec. 

So far, only CameraSensorSpec exists, but I'm adding the structure for checks so that when Fisheye/Equirectangular sensors are implemented, we can follow the logic to initialize the right sensorSpec.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 4 above)
Comment (vauduong) [reply habitat/sims/habitat_simulator/habitat_simulator.py:266]: Are you saying that instead of checking if it's in a set we should get the value it and compare it with the enum numbers? This is just a way to check which type of SensorSpec to initialize based on the values in the config.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 1 above)
Comment (mathfac) [reply habitat/sims/habitat_simulator/habitat_simulator.py:294]: Yes, we have only [CollisionMeasure](https://github.com/facebookresearch/habitat-lab/blob/aa24551fd9942b11b7af4a076d9c05e080594d08/habitat/tasks/nav/nav.py#L670), but that would be great to have BumpSensor, but it looks like it can be HabitatLab Task sensor, not Simulator one.

**Comment by mathfac** [comment]
Looking forward for this change.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 3 above)
Comment (mathfac) [reply habitat/sims/habitat_simulator/habitat_simulator.py:167]: Does Subtype infer how observation space will be defined? Does different Subtypes of sensors require separate classes wrappers in HabLab?

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Want to let @erikwijmans to take a look at this before merging. Otherwise, looks good to me. We may want to add an error to upgrade habitat_sim though.

File: habitat/sims/habitat_simulator/habitat_simulator.py (hunk 3 above)
Comment (vauduong) [reply habitat/sims/habitat_simulator/habitat_simulator.py:167]: Hi @mathfac , I think I don't understand enough about how observation space is defined in hablab vs habsim. What is the difference there and what should I be aware of?

Comment (Skylion007) [reply habitat/sims/habitat_simulator/habitat_simulator.py:167]: Exactly, I don't think they should require separate wrappers, they should be set by the config.

**Review by erikwijmans — APPROVED** [review]
This looks fine to me!
